File: problems/77/solution.cpp
// Accepted: 104ms (85.34%), 81.50MB (48.01%)

class Solution {
    vector<vector<int>> combinations;
    vector<int> generated;

    void go(int begin, int n, int k) {
        if (k == 0) {
            combinations.push_back(generated);
            return;
        }

        for (int i = begin; i < n-k+2; ++i) {
            generated.push_back(i);
            go(i+1, n, k-1);
            generated.pop_back();
        }
    }
public:
    vector<vector<int>> combine(int n, int k) {
        go(1, n, k);

        return combinations;
    }
};

File: problems/77/solution.ts
// Accepted: 465ms (61.87%), 121.22MB (98.44%)

function combine(n: number, k: number): number[][] {
    const combinations: number[][] = [];

    const generated: number[] = [];
    const go = (begin: number, k: number) => {
        if (k === 0) {
            combinations.push(generated.slice(0));
            return;
        }

        for (let i = begin; i < n-k+2; ++i) {
            generated.push(i);
            go(i+1, k-1);
            generated.pop();
        }
    };
    go(1, k);

    return combinations;
};
